[basil-boh] iP

basil-boh opened the pull request:
### Steve 
 

> ~ _Do-do-do do do-do do, your friend....Steve_! - Rent-free Tiktok Sound Track 🔥

**Steve is your all in one friend that keeps track of tasks for you! 🫵🏻** 
- Intuitive commands ✅
- Available 24/7 🕔
- Lighting Fast ⚡

**To get bad boy Steve going, you need to:**

1. Download it from here
2. Double click it
3. Use by entering commands followed by task description
4.  Leave the rest to Steve ❤️

So...what are you waiting for? Try Steve out for free!!! 🤓

Features:

- [ ] 1. To do tasks 
- [ ] 2. Deadline tasks
- [ ] 3. Event Tasks
- [ ] 4. List
- [ ] 5. Mark/Unmark
- [ ] 6. Delete

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```
package steve.ui;

import steve.tasks.Task;
import steve.tasks.TaskManager;
import steve.ui.UserInterface;
import steve.storage.FileHandler;
import java.util.ArrayList;

// Main Class File
public class Steve{
    public static void main(String[] args) {
        FileHandler fileHandler = new FileHandler("src/main/data/steve.txt");
        ArrayList<Task> tasks = fileHandler.loadTasks(); // Load tasks from file
        TaskManager taskManager = new TaskManager(tasks); // Initialize TaskManager with loaded tasks
        UserInterface ui = new UserInterface(taskManager); // Initialize UserInterface with TaskManager

        // Start the application
        ui.start();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

basil-boh committed:
- Renamed main class to Steve
- Added simple static Greeting

basil-boh committed:
- Added user input
- Steve able to echo user inputs
- Exits on 'bye' command

basil-boh committed:
- Added userData array storing up to 100 tasks
- Returns list of tasks with command 'list'

basil-boh committed:
- Added Task class
- Modified userData from string array to task object array
- Added mark task functionality
- Added unmark task functionality

basil-boh committed:
- Created Event, Deadline, ToDo classes
- Reformatted main class to utilise more OOP

basil-boh committed:
- Added Exception Handling for cases:
1. InvalidTaskNumberException
2. NumberFormatException
3. emptyDescription

- Added testcases for expected I/O

basil-boh committed:
- Added Delete feature
- Modified static array to dynamic array list

basil-boh committed:
- improved code reusability in Task related files

basil-boh committed:
- Added Enum

basil-boh committed:
- Improved code reusability

basil-boh committed:
- Added comments for readability

basil-boh committed:
- Correct comments

basil-boh committed:
- Saves tasks to steve.txt file

basil-boh committed:
- Able to successfully load in tasks

basil-boh committed:
- Added comment

basil-boh committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  - Able to successfully load in tasks
  - Saves tasks to steve.txt file

basil-boh committed:
- Added date and time parsing for the deadline class

basil-boh committed:
- Added date and time parsing for the event class

basil-boh committed:
- Added merge commit comment

basil-boh committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  - Added date and time parsing for the event class
  - Added date and time parsing for the deadline class

basil-boh committed:
- Refactored main class with addition of more OOP
- Added UI Class
- Added TaskManager Class
- Added FileHandler
- Added CommandCreate Classes: Specifically for parsing and creating diff commands

basil-boh committed:
- Added comment

basil-boh committed:
Merge branch 'branch-A-MoreOOP'

* branch-A-MoreOOP:
  - Refactored main class with addition of more OOP - Added UI Class - Added TaskManager Class - Added FileHandler - Added CommandCreate Classes: Specifically for parsing and creating diff commands

basil-boh committed:
- Created respective packages

basil-boh committed:
- Created respective packages pt2

basil-boh committed:
Merge commit '81a9c537afe8492675eb253968939364d65db091'

* commit '81a9c537afe8492675eb253968939364d65db091':
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

basil-boh committed:
- Gradle setup

basil-boh committed:
- Added JUnit testing for ToDo and Event classes

basil-boh committed:
- Added header comments

basil-boh committed:
- Improved coding style

basil-boh committed:
- Implemented Find
- Added FindCommand
- Amended CommandCreate and TaskManager

basil-boh committed:
Merge branch 'branch-A-Coding-Standard'

* branch-A-Coding-Standard:
  - Improved coding style

basil-boh committed:
- Amended merge conflicts

JohannsenLum commented on src/main/java/steve.tasks/Task.java:
```diff
@@ -0,0 +1,118 @@
+public class Task {
+    protected String description;
+    protected boolean done;
+    protected static int taskCount = 0;
```
I'm not sure if your static taskCount would be still needed as you have your TaskManager to handle the list of tasks

JohannsenLum reviewed (commented):
Your code is well-structured and follows good practices!

Jovin-Ang commented on build.gradle:
```diff
@@ -0,0 +1,44 @@
+    testRuntimeOnly group: 'org.junit.jupiter', name: 'junit-jupiter-engine', version: '5.10.0'
+    implementation group: 'com.joestelmach', name: 'natty', version: '0.13'
+    testImplementation group: 'org.junit.jupiter', name: 'junit-jupiter-api', version: '5.10.0'
+    testRuntimeOnly group: 'org.junit.jupiter', name: 'junit-jupiter-engine', version: '5.10.0'
```
Opps :sweat_smile:, I think this looks like duplicate lines

basil-boh replied:
Thanks for pointing that out!

Jovin-Ang commented on src/main/java/steve.commands/Command.java:
```diff
@@ -0,0 +1,7 @@
+package steve.commands;
+import steve.exceptions.InvalidTaskNumberException;
```
Perhaps there should be an empty line in between here. I noticed this in `ListCommand.java` too.

basil-boh replied:
thanks for pointing that out!

Jovin-Ang commented on src/main/java/steve.commands/CommandCreate.java:
```diff
@@ -0,0 +1,32 @@
+                return new DeleteCommand(taskManager, userInput);
+            default:
+                throw new InvalidCommandException("Unknown command: " + firstWord);
+        }
```
I believe that the case clauses should have **no** indentation according to the coding standard

basil-boh replied:
You're right will take note of this

Jovin-Ang commented on src/main/java/steve.commands/MarkCommand.java:
```diff
@@ -0,0 +1,23 @@
+        taskManager.markTask(taskNumber, true); // Mark as done
+    }
+
+
```
Are these extra line breaks intended?

basil-boh replied:
Good catch thanks!

Jovin-Ang commented on src/main/java/steve.tasks/Deadline.java:
```diff
@@ -0,0 +1,87 @@
+            this.description = "Description cannot be empty. Usage: deadline <description /by yyyy-MM-dd HHmm>";
+            super.decreaseTaskCount();
+        }
+    }
```
This part is a little unclear to me. Is this error message supposed to be displayed to the user or stored as the task description? Perhaps you could add some code comments?
Since this is a constructor, the Deadline object will still be created provided an exception is thrown. Perhaps you could extract the checking logic elsewhere and only pass valid parameters to the constructor.
Similar comment for the other types of tasks.

Jovin-Ang commented on src/main/java/steve.tasks/Task.java:
```diff
@@ -0,0 +1,118 @@
+            task.message(firstWord);
+        }
+    }
+    public static void validateTaskNumberDel(int taskNumber, int taskCount, ArrayList<Task> userData) throws InvalidTaskNumberException {
```
Missing empty line in between

Jovin-Ang commented on src/main/java/steve/ui/Messages.java:
```diff
@@ -0,0 +1,35 @@
+        String first = "Sorry I don't quite understand!";
+        String second = " Please add a task by entering: \n 1. todo \n 2. deadline \n 3. event \n 4. find \n";
+        String third = "Followed by a description of your task! =)";
+        System.out.println(first + second + third);
```
Perhaps you can condense this into a single multi line string. Likewise for several other cases.
```suggestion
        System.out.println("""
        Sorry I don't quite understand! Please add a task by entering: 
        1. todo
        2. deadline
        3. event
        4. find
        Followed by a description of your task! =)""");
```

basil-boh replied:
Great suggestion will adopt it

Jovin-Ang commented on src/main/java/steve.tasks/Deadline.java:
```diff
@@ -0,0 +1,87 @@
+
+// Deadline task
+public class Deadline extends Task {
+    protected String description;
```
There is no need to redeclare a instance variable that is already present in the superclass.

Jovin-Ang reviewed (commented):
I noticed the task classes are a little long. Perhaps it can be further simplified by removing some unnecessary methods which are not used and placing common instance variables into the `Task.java` superclass, such as `isValid`, which seems to be present in all 3 subclasses.

GrassHeadd commented on build.gradle:
```diff
@@ -0,0 +1,44 @@
+    testRuntimeOnly group: 'org.junit.jupiter', name: 'junit-jupiter-engine', version: '5.10.0'
+    implementation group: 'com.joestelmach', name: 'natty', version: '0.13'
+    testImplementation group: 'org.junit.jupiter', name: 'junit-jupiter-api', version: '5.10.0'
+    testRuntimeOnly group: 'org.junit.jupiter', name: 'junit-jupiter-engine', version: '5.10.0'
```
As mentioned by Jovin, I think there are duplicated lines, I assume you probably copied the tutorial's 2 lines below much as I did, I think maybe can review more carefully next time to see that its mostly already in the original repo!

GrassHeadd commented on src/main/data/steve.txt:
actually this file could be just added to gitignore since given an actual todo chatbot, you would probably not want to have your own txt files be kept in others' computer, would help to just add this line
```
steve.txt
```
to tell github to ignore the changes you made locally to this file

basil-boh replied:
rly good idea I completely forgot about gitignore haha

GrassHeadd commented on src/main/java/steve.commands/Command.java:
```diff
@@ -0,0 +1,7 @@
+package steve.commands;
+import steve.exceptions.InvalidTaskNumberException;
```
as mentioned by jovin, from my vague memory of 2030s and just general java programming, the general consensus for this scenario is to separate the package line and imports with an empty line in between

GrassHeadd commented on src/main/java/steve.commands/Command.java:
just wanted to add that the commands file is really well abstracted!

basil-boh replied:
thank you :)

GrassHeadd commented on src/main/java/steve.commands/FindCommand.java:
just curious about why you didnt add a description to this class, also i believe we are suppose to add javadocs to the whole class too

basil-boh replied:
truthfully I went for a break and forgot to add it afterwards hahaha but will def add it in soon

GrassHeadd commented on src/main/java/steve.tasks/Deadline.java:
```diff
@@ -0,0 +1,87 @@
+
+    @Override
+    public String getOriginalDescription() {
+        return this.description + " /by " + (this.by != null ? this.by.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm")) : "");
```
can split this into 2 lines since its kind of long

basil-boh replied:
Agreed, will do

GrassHeadd commented on src/main/java/steve.tasks/Task.java:
```diff
@@ -0,0 +1,118 @@
+        System.out.println(result);
+    }
+
+    public void markDoneOrNot(boolean isDone) {
+        this.done = isDone;
+    }
```
name feels unconventional, maybe could have put toggleDone or smth

GrassHeadd commented on src/main/java/steve.tasks/Task.java:
```diff
@@ -0,0 +1,118 @@
+        taskCount--;
+    }
+
+    public String list() {
+        return "." + " [" + this.code() + "]" + "[" + this.getStatusIcon() + "] " + this.getDescription();
+    }
```
i feel listTask is a better name but list is fine too!

GrassHeadd commented on src/main/java/steve.tasks/ToDo.java:
```diff
@@ -0,0 +1,65 @@
+    protected String description;
+    protected boolean isValid = false;
+
+    public ToDo(String description) {
+        super(description == null || description.trim().isEmpty()
+                ? "Description cannot be empty. Usage: todo <description>"
+                : description);
```
can maybe stop them from even initialising the tasks by using a try catch block rather than letting the task classes handle it

GrassHeadd commented on src/main/java/steve/storage/FileHandler.java:
```diff
@@ -0,0 +1,51 @@
+
+    private static Task parseTask(String line) {
+        String firstWord = line.split(":")[0];
+        String restOfLine = line.substring(firstWord.length() + 1).trim();
```
continuation of the above comment, maybe can put smth like but honestly i think your implementation is fine, just an alternative i guess that i personally would do
```
try {
  String restOfLine = line.substring(firstWord.length() + 1).trim();
} catch (NullPointerException e) {
  System.out.println("Description cannot be empty.")
}
```

GrassHeadd reviewed (commented):
Overall really well done! Note worthy things are the good abstractions of the different classes and packages, also created your own exceptions, however does not seem to add javadocs to the files and could have utilised the built in exceptions better. Great work!

basil-boh committed:
- Add javadoc comments for all classes and methods
- amend code to adhere to checkstyle

basil-boh committed:
- Configure javafx files
- Create first node, scene & stage

basil-boh committed:
- Add graphics and echo

basil-boh committed:
- Integrate GUI with Steve class logics

basil-boh committed:
Fix checkstyle errors

Find.java: only has one long method to filter out tasks

filterString() method is handling too many responsibilities

Add several methods to handle specific responsibilities

improves code readability and modularity

basil-boh committed:
improve coding styles for todo, task and taskmanager classes

fix save tasks feature

basil-boh committed:
delete task does not save correctly

task will reappear in task list on chatbot reboot

fix delete and save functionality in Task.java and TaskManager.java

fix delete bug where incorrect amount of tasks are shown

refactored event.java class to improve readability and reusability

basil-boh committed:
refactor Deadline.java class to improve readability and reusability

basil-boh committed:
add contact.java class to enter contact names

basil-boh committed:
implement client feature

basil-boh committed:
improve UI

basil-boh committed:
implement personality by adjusting message lines

basil-boh committed:
improve error handling

basil-boh committed:
add more JUnit testing

basil-boh committed:
update README.md

basil-boh committed:
Update README.md

basil-boh committed:
jar file unable to save

fix jar saving bug by adding a create steve.txt file if not found

basil-boh committed:
Amend TaskManager comment

basil-boh committed:
Merge branch 'branch-A-Level-10'

basil-boh committed:
Amended TaskManager.java addTask comment

basil-boh committed:
Merge branch 'branch-A-Assertions'

basil-boh committed:
Remove comment from Steve.java

basil-boh committed:
Merge branch 'branch-A-CodeQuality' Resolve merge conflicts in Steve.java

basil-boh committed:
Update README.md display details

basil-boh committed:
Update README.md marking details

basil-boh committed:
Update README.md unmark details

basil-boh committed:
Update README.md find details

basil-boh committed:
Update README.md client details
